pkg/gardenlet/controller/gardenlet/reconciler.go

```diff
@@ -0,0 +1,191 @@
+	status.ObservedGeneration = gardenlet.Generation
+
+	_, gardenletConfig, err := helper.ExtractSeedTemplateAndGardenletConfig(gardenlet.Name, &gardenlet.Spec.Config)
+	if err != nil {
+		r.Recorder.Eventf(gardenlet, corev1.EventTypeWarning, gardencorev1beta1.EventReconcileError, err.Error())
+		updateCondition(r.Clock, status, gardencorev1beta1.ConditionFalse, gardencorev1beta1.EventReconcileError, err.Error())
+		if updateErr := r.updateStatus(ctx, gardenlet, status); updateErr != nil {
+			log.Error(updateErr, "Could not update status")
+		}
+		return reconcile.Result{}, fmt.Errorf("error extracting gardenlet configuration: %w", err)
+	}
+
+	seed, err := managedseed.GetSeed(ctx, r.GardenClient, gardenlet.Name)
+	if err != nil {
+		r.Recorder.Eventf(gardenlet, corev1.EventTypeWarning, gardencorev1beta1.EventReconcileError, err.Error())
+		updateCondition(r.Clock, status, gardencorev1beta1.ConditionFalse, gardencorev1beta1.EventReconcileError, err.Error())
+		if updateErr := r.updateStatus(ctx, gardenlet, status); updateErr != nil {
+			log.Error(updateErr, "Could not update status")
+		}
+		return reconcile.Result{}, fmt.Errorf("error getting seed: %w", err)
+	}
+
+	log.Info("Deploying gardenlet")
+	r.Recorder.Eventf(gardenlet, corev1.EventTypeNormal, gardencorev1beta1.EventReconciling, "Deploying gardenlet")
+	if err := r.deployGardenlet(ctx, log, gardenlet, seed, gardenletConfig); err != nil {
+		r.Recorder.Eventf(gardenlet, corev1.EventTypeWarning, gardencorev1beta1.EventReconcileError, err.Error())
+		updateCondition(r.Clock, status, gardencorev1beta1.ConditionFalse, gardencorev1beta1.EventReconcileError, err.Error())
+		if updateErr := r.updateStatus(ctx, gardenlet, status); updateErr != nil {
+			log.Error(updateErr, "Could not update status")
+		}
+		return reconcile.Result{}, fmt.Errorf("error deploying gardenlet: %w", err)
+	}
+
+	log.V(1).Info("Reconciliation finished")
+	r.Recorder.Eventf(gardenlet, corev1.EventTypeNormal, gardencorev1beta1.EventReconciled, "Gardenlet has been deployed")
+	updateCondition(r.Clock, status, gardencorev1beta1.ConditionTrue, gardencorev1beta1.EventReconciled, "Gardenlet with chart from "+gardenlet.Spec.Deployment.Helm.OCIRepository.GetURL()+" has been deployed")
+	if updateErr := r.updateStatus(ctx, gardenlet, status); updateErr != nil {
+		log.Error(updateErr, "Could not update status")
+	}
+
+	return reconcile.Result{RequeueAfter: r.Config.Controllers.Gardenlet.SyncPeriod.Duration}, nil
+}
+
+func (r *Reconciler) deployGardenlet(
+	ctx context.Context,
+	log logr.Logger,
+	gardenlet *seedmanagementv1alpha1.Gardenlet,
+	seed *gardencorev1beta1.Seed,
+	gardenletConfig *gardenletv1alpha1.GardenletConfiguration,
+) error {
+	values, err := r.prepareGardenletChartValues(ctx, log, gardenlet, seed, gardenletConfig)
+	if err != nil {
+		return fmt.Errorf("failed preparing gardenlet chart values: %w", err)
+	}
+
+	archive, err := r.HelmRegistry.Pull(ctx, &gardenlet.Spec.Deployment.Helm.OCIRepository)
```

Comment:
The gardenlet helm chart should be limited in size. However, what happens if due to operator failure or malicious intent a big image is configured here. Will this effectively cause a denial of service in gardenlet as the full image is read into memory leading potentially to an out-of-memory error? This could lead to a death loop as the update would probably retried on start-up.

Reply:
Yes, it will. However, there are many cases like this in resources that can only be controlled via Gardener operators. I don't think it makes sense to consider such intentionally malicious configurations explicitly. In order to get `gardenlet` into a death loop, it is not even necessary to upload a malicious image into the OCI repo. Just misconfigure its config in the `Gardenlet`'s `.spec.config` - way easier. :)